prow/cmd/exporter/main.go

```diff
@@ -0,0 +1,128 @@
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"k8s.io/test-infra/pkg/flagutil"
+	prowjobinformer "k8s.io/test-infra/prow/client/informers/externalversions"
+	"k8s.io/test-infra/prow/config"
+	prowflagutil "k8s.io/test-infra/prow/flagutil"
+	"k8s.io/test-infra/prow/interrupts"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/metrics"
+	"k8s.io/test-infra/prow/pjutil"
+)
+
+type options struct {
+	configPath     string
+	jobConfigPath  string
+	selectorString string
+
+	kubernetes prowflagutil.KubernetesOptions
+
+	selector labels.Selector
+}
+
+func gatherOptions(fs *flag.FlagSet, args ...string) options {
+	var o options
+
+	fs.StringVar(&o.configPath, "config-path", "", "Path to config.yaml.")
+	fs.StringVar(&o.jobConfigPath, "job-config-path", "", "Path to prow job configs.")
+	fs.StringVar(&o.selectorString, "label-selector", labels.Everything().String(), "Label selector to be applied in prowjobs. See https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors for constructing a label selector.")
+
+	for _, group := range []flagutil.OptionGroup{&o.kubernetes} {
+		group.AddFlags(fs)
+	}
+	if err := fs.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatalf("cannot parse args: '%s'", os.Args[1:])
+	}
+	o.configPath = config.ConfigPath(o.configPath)
+	return o
+}
+
+func (o *options) Validate() error {
+	for _, group := range []flagutil.OptionGroup{&o.kubernetes} {
+		if err := group.Validate(false); err != nil {
+			return err
+		}
+	}
+
+	selector, err := labels.Parse(o.selectorString)
+	if err != nil {
+		return fmt.Errorf("parse label selector: %v", err)
+	}
+	o.selector = selector
+	return nil
+}
+
+func mustRegister(component string, selector labels.Selector, lister lister) *prometheus.Registry {
+	registry := prometheus.NewRegistry()
+	prometheus.WrapRegistererWith(prometheus.Labels{"collector_name": component}, registry).MustRegister(&prowJobCollector{
+		name:     component,
+		lister:   lister,
+		selector: selector,
+	})
+	registry.MustRegister(
+		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
+		prometheus.NewGoCollector(),
+	)
+	return registry
+}
+
+func main() {
+	logrusutil.ComponentInit("exporter")
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	if err := o.Validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
+	}
+
+	defer interrupts.WaitForGracefulShutdown()
+
+	pjutil.ServePProf()
+
+	configAgent := &config.Agent{}
+	if err := configAgent.Start(o.configPath, o.jobConfigPath); err != nil {
+		logrus.WithError(err).Fatal("Error starting config agent.")
+	}
+	cfg := configAgent.Config
+
+	pjClientset, err := o.kubernetes.ProwJobClientset(cfg().ProwJobNamespace, false)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create prowjob client set")
+	}
+	informerFactory := prowjobinformer.NewSharedInformerFactory(pjClientset, 15*time.Second)
```

Comment:
Any reason not to use the `controller-runtime` client like e.G. [here](https://github.com/kubernetes/test-infra/blob/35d5f5ae4abf35339c23d10ea6cca1f2c994a577/prow/cmd/sinker/main.go#L131-L153)?

It has the advantage that it can be used for any type of resource and not just prowjobs. The implementation you get from the `mgr` is also cache-backed.

Reply:
Checking ...
No special reason ... just i started with plank, not sinker ... :)

Reply:
Checked.

I think I prefer to keeping this one.
3 lines of code to get a lister is more succinct (as long as it works ... i have to admit that I have not tested a lister from informer before while `grep` showed me similar [code exists in prow](https://github.com/kubernetes/test-infra/blob/35d5f5ae4abf35339c23d10ea6cca1f2c994a577/prow/cmd/crier/main.go#L250) already) than `controller-runtime` and we do not need CRDs other than prow-jobs, at least for the moment.
We can switch if it changes in the future.